### PR TITLE
impl: add client origin metadata for a failed async unary rpc

### PR DESCRIPTION
**Pull request by alevenberg**
https://github.com/googleapis/google-cloud-cpp/issues/13724

This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14088)

### CODE REVIEW

File: google/cloud/internal/async_rpc_details.h [hunk 1]
```diff
@@ -73,7 +74,8 @@ class AsyncUnaryRpcFuture : public AsyncGrpcOperation {
     if (!ok) {
       // `Finish()` always returns `true` for unary RPCs, so the only time we
       // get `!ok` is after `Shutdown()` was called; treat that as "cancelled".
-      promise_.set_value(Status(StatusCode::kCancelled, "call cancelled"));
+      promise_.set_value(internal::CancelledError("gRPC async unary RPC failed",
```
Comment (dbolduc) [inline google/cloud/internal/async_rpc_details.h:77]: Why? The original message is more accurate?

File: google/cloud/internal/async_rpc_details.h [hunk 2]
```diff
@@ -73,7 +74,8 @@ class AsyncUnaryRpcFuture : public AsyncGrpcOperation {
...
+                                                  GCP_ERROR_INFO()));
```
Comment (dbolduc) [inline google/cloud/internal/async_rpc_details.h:78]: I have not used this, but we would want something like:

`GCP_ERROR_INFO().WithMetadata("gl-cpp.error.origin", "client")`

https://github.com/googleapis/google-cloud-cpp/blob/226c607e572d43d35647f2a07f453ca1a00d19b9/google/cloud/internal/rest_stub_helpers.cc#L41

**Review by dbolduc — COMMENTED** [review]
`internal` isn't really a component. So we should just say `impl: ...`

File: google/cloud/internal/async_rpc_details.h (hunk 1 above)
Comment (alevenberg) [reply google/cloud/internal/async_rpc_details.h:77]: I thought the original message was ambiguous. I can revert to that.

File: google/cloud/internal/async_rpc_details.h (hunk 2 above)
Comment (alevenberg) [reply google/cloud/internal/async_rpc_details.h]: Thanks for the pointer. Done.

Added a test too.

File: google/cloud/internal/async_rpc_details.h (hunk 1 above)
Comment (alevenberg) [reply google/cloud/internal/async_rpc_details.h]: Reverted the message.

File: google/cloud/internal/async_rpc_details.h [hunk 3]
```diff
@@ -72,8 +73,14 @@ class AsyncUnaryRpcFuture : public AsyncGrpcOperation {
     ScopedCallContext scope(call_context_);
     if (!ok) {
       // `Finish()` always returns `true` for unary RPCs, so the only time we
-      // get `!ok` is after `Shutdown()` was called; treat that as "cancelled".
-      promise_.set_value(Status(StatusCode::kCancelled, "call cancelled"));
+      // get `!ok` is after `Shutdown()` was called; create a client-origin
+      // "cancelled". By marking the status, we can use that to determine
+      // whether it *might* not be safe to make a GetServerInitialMetadata call
+      // on the client context. This omits the grpc server initial metadata
+      // tracing information from the span attributes.
```
Comment (devbww) [inline google/cloud/internal/async_rpc_details.h:80]: I'd probably leave the GetServerInitialMetadata()/span-attributes part of this comment to the code that checks for the "origin=client".

Comment (alevenberg) [reply google/cloud/internal/async_rpc_details.h]: Done

File: google/cloud/completion_queue_test.cc [hunk 4]
```diff
@@ -404,7 +405,11 @@ TEST(CompletionQueueTest, MakeRpcsAfterShutdown) {
           },
           r1, std::make_unique<grpc::ClientContext>())
           .then([](future<StatusOr<Response>> f) {
-            EXPECT_EQ(StatusCode::kCancelled, f.get().status().code());
+            auto status = f.get().status();
+            EXPECT_EQ(StatusCode::kCancelled, status.code());
```
Comment (devbww) [inline google/cloud/completion_queue_test.cc:409]: Nit: Consider `EXPECT_THAT(status, StatusIs(StatusCode::kCancelled));` for a better error message.

**Review by devbww — APPROVED** [review]

File: google/cloud/completion_queue_test.cc (hunk 4 above)
Comment (alevenberg) [reply google/cloud/completion_queue_test.cc]: done

File: google/cloud/internal/async_rpc_details.h (hunk 3 above)
Comment (alevenberg) [reply google/cloud/internal/async_rpc_details.h]: done

**Pull request merged**